### PR TITLE
fix: WebAuthenticationBroker returns canonical URL with the trailing

### LINKED ISSUE #6485: WebAuthenticationBroker returns canonical URL with the trailing slash which is not handled by Android
``` C#
[Activity(Label = "MsalActivity")]
[IntentFilter(
    new[] { Intent.ActionView },
    Categories = new[] { Intent.CategoryBrowsable, Intent.CategoryDefault },
    DataHost = "auth/",
    DataScheme = Constants.AppScheme)]
public class MsalActivity : BrowserTabActivity { }
```
If the trailing slash is specified in `IntentFilter` Android cannot handle the scheme redirect.

## Current behavior

`WebAuthenticationBroker.GetCurrentApplicationCallbackUri()` does return trailing slash for Android

## Expected behavior

`WebAuthenticationBroker.GetCurrentApplicationCallbackUri()` does not return trailing slash for Android

## How to reproduce it (as minimally and precisely as possible)
1. Take WASM sample
2. Set in the constructor of the app.xam.cs
``` C#
#if __WASM__
            WinRTFeatureConfiguration.WebAuthenticationBroker.DefaultCallbackPath = "/authentication/login-callback.htm";
#elif !WINDOWS_UWP
            WinRTFeatureConfiguration.WebAuthenticationBroker.DefaultReturnUri = new Uri($"{Constants.AppScheme}://{Constants.AppHost}");
#endif
```
3. Add the trailing slash to
https://github.com/unoplatform/Uno.MSAL.Graph.Demo/blob/1aaa92eecd25d1e36c782711c54f7963d5ca0abd/Uno.MSAL.Graph.Demo/Uno.MSAL.Graph.Demo.Droid/MsalActivity.cs#L10
4. Put `WebAuthenticationBroker.GetCurrentApplicationCallbackUri()` instead of
https://github.com/unoplatform/Uno.MSAL.Graph.Demo/blob/1aaa92eecd25d1e36c782711c54f7963d5ca0abd/Uno.MSAL.Graph.Demo/Uno.MSAL.Graph.Demo.Shared/MainPage.xaml.cs#L44

## Workaround

Set redirect URI manually from constant

## Environment

Nuget Package:
- Uno.UI

Nuget Package Version(s): 3.8.13

Affected platform(s):

- Android

### PR DESCRIPTION
**Pull request by aesteves900**
…slash which is not handled by Android

GitHub Issue (If applicable): closes #6485

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

WebAuthenticationBroker.GetCurrentApplicationCallbackUri() does return trailing slash for Android

## What is the new behavior?

WebAuthenticationBroker.GetCurrentApplicationCallbackUri() does not return trailing slash for Android

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ x] Validated PR `Screenshots Compare Test Run` results.
- [ x] Contains **NO** breaking changes
- [ x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs [hunk 1]
```diff
@@ -33,6 +33,7 @@ internal static partial class StringExtensions
 #else
 		private static readonly Lazy<Regex> _newLineRegex = new Lazy<Regex>(() => new Regex(@"^", RegexOptions.Multiline));
 #endif
+		public static Uri TrimEndUriSlash(this Uri uri) => uri.ToString().EndsWith('/') ? new Uri(uri.ToString().TrimEnd(new[] { '/' })) : uri;
```
Comment (xperiandri) [inline src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs:36]: I suppose that it is more resource-efficient to call `uri.AbsolutePath.EndsWith('/')` then materializing the whole `Uri` into a `string`

I propose such changes
``` C#
public static Uri TrimEndUriSlash(this Uri uri) => uri.AbsolutePath.EndsWith('/') ? new Uri(uri.ToString()[..^1]) : uri;
```

**Review by xperiandri — CHANGES_REQUESTED** [review]

File: src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs (hunk 1 above)
Comment (xperiandri) [inline src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs:36]: I suppose that it is more resource-efficient to call `uri.AbsolutePath.EndsWith('/')` then materializing the whole `Uri` into a `string`

I propose such changes

```suggestion
		public static Uri TrimEndUriSlash(this Uri uri) => uri.AbsolutePath.EndsWith('/') ? new Uri(uri.ToString()[..^1]) : uri;
```

**Review by xperiandri — CHANGES_REQUESTED** [review]

File: src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs (hunk 1 above)
Comment (aesteves900) [reply src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs:36]: Nice @xperiandri, I commi your suggestion, thnak you!

**Review by MartinZikmund — APPROVED** [review]

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs [hunk 2]
```diff
@@ -42,7 +48,12 @@ public Uri GetCurrentApplicationCallbackUri()
 				this.Log().Warn(message);
 			}
 
+#if __ANDROID__
+
```
Comment (MartinZikmund) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs:52]: Please add a comment with explanation why the trim is needed

**Comment by xperiandri** [comment]
@MartinZikmund #6485

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs (hunk 2 above)
Comment (xperiandri) [reply src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs:52]: #6485
Do you mean comment from some Android docs?

Comment (MartinZikmund) [reply src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs:52]: Either that or just a note that the trim is necessary because otherwise it fails. Or maybe even split it like this:

```c#
var uri = new Uri(schemes[0] + WinRTFeatureConfiguration.WebAuthenticationBroker.DefaultCallbackPath);
#if __ANDROID__
uri = uri.TrimEndUriSlash()
#endif
```

To avoid duplication

File: src/Uno.Foundation/UriExtensions.cs [hunk 3]
```diff
@@ -26,6 +26,10 @@ public static class UriExtensions
 		}
 
 
+#if __ANDROID__
+		public static Uri TrimEndUriSlash(this Uri uri) => uri.AbsolutePath.EndsWith("/") ? new Uri(uri.ToString()[..^1]) : uri;
```
Comment (jeromelaban) [inline src/Uno.Foundation/UriExtensions.cs:30]: Let's place this method internal for now. We'll see later if we need to make it public, as needed.

Comment (aesteves900) [reply src/Uno.Foundation/UriExtensions.cs:30]: If the method becomes internal will be broken unit tests and webAuthenticationBroker.

Comment (carldebilly) [reply src/Uno.Foundation/UriExtensions.cs:30]: Should work as internal. There's an `[assembly:InternalsVisibleTo]` attribute for that.

File: src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs [hunk 4]
```diff
@@ -34,6 +34,10 @@ internal static partial class StringExtensions
 		private static readonly Lazy<Regex> _newLineRegex = new Lazy<Regex>(() => new Regex(@"^", RegexOptions.Multiline));
 #endif
 
+#if __ANDROID__
+		public static Uri TrimEndUriSlash(this Uri uri) => uri.AbsolutePath.EndsWith("/") ? new Uri(uri.ToString()[..^1]) : uri;
+#endif
```
[the remark this answers is not in the document]
Comment (carldebilly) [reply src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs]: The change should not be Android-specific

[the remark this answers is not in the document]
Comment (aesteves900) [reply src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs]: @carldebilly , the problem was reported just to Android, because of this I put the condition. Can I increase the scope for all the platforms?

[the remark this answers is not in the document]
Comment (carldebilly) [reply src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs]: There's a "duplication risk" when it's platform specific.  Anyway it's getting linked-out during compilation when it's not used, so it's exactly a problem.

TL;DR: Yes it should be everywhere.

File: src/Uno.Foundation/UriExtensions.cs (hunk 3 above)
Comment (aesteves900) [reply src/Uno.Foundation/UriExtensions.cs]: Done!

File: src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (carldebilly) [reply src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs]: Please wait for the changes to be pushed before clicking on the "Resolve conversation" button.

**Review by carldebilly — CHANGES_REQUESTED** [review]
Please wait for the changes to be pushed before resolving conversations... it's easier to follow.

File: src/Uno.Foundation/UriExtensions.cs [hunk 5]
```diff
@@ -26,6 +26,10 @@ public static class UriExtensions
 		}
 
 
+#if __ANDROID__
+		internal static Uri TrimEndUriSlash(this Uri uri) => uri.AbsolutePath.EndsWith("/") ? new Uri(uri.ToString()[..^1]) : uri;
```
Comment (jeromelaban) [inline src/Uno.Foundation/UriExtensions.cs:30]: `.ToString()` and `OriginalString` behave differently, particularly when dealing with spaces or other similar escaped characters. 

I'd rather see this:
```suggestion
		internal static Uri TrimEndUriSlash(this Uri uri) => uri.OriginalString.TrimEnd('/');
```

Comment (aesteves900) [reply src/Uno.Foundation/UriExtensions.cs:30]: @jeromelaban , this will create a new instance from Uri all the time. Any concern about that? 

`internal static Uri TrimEndUriSlash(this Uri uri) => uri.OriginalString.TrimEnd('/');`

`internal static Uri TrimEndUriSlash(this Uri uri) => new Uri(uri.OriginalString.TrimEnd('/'));`

File: src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (aesteves900) [reply src/Uno.Foundation/Uno.Core.Extensions/Uno.Core.Extensions/StringExtensions.cs]: I removed. Thank you!

File: src/Uno.Foundation/UriExtensions.cs (hunk 5 above)
Comment (jeromelaban) [reply src/Uno.Foundation/UriExtensions.cs]: Even if it's a low-traffic area, that's true indeed, thanks. The `OriginalString` comment is still valid, though.

Comment (aesteves900) [reply src/Uno.Foundation/UriExtensions.cs]: I did the recomended change. Thank you!

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs [hunk 6]
```diff
@@ -21,7 +21,13 @@ public Uri GetCurrentApplicationCallbackUri()
 			{
 				// A default one is defined: we're using it instead of using
 				// a discovery mechanism.
+
+#if __ANDROID__
+				return defaultUri.TrimEndUriSlash();
```
Comment (carldebilly) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs:26]: Still need to explain in the code why we need this on Android.

File: src/Uno.UI.RuntimeTests/Tests/Extensions/Given_Uri.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+using Windows.Web.Http;
+
+namespace Uno.UI.RuntimeTests.Tests.Extensions
+{
+	[TestClass]
+	public class Given_Uri
+	{
+
+#if __ANDROID__
+		[TestMethod]
```
Comment (carldebilly) [inline src/Uno.UI.RuntimeTests/Tests/Extensions/Given_Uri.cs:13]:
```suggestion
		[TestMethod]
```

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs (hunk 6 above)
Comment (aesteves900) [reply src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs:26]: It was removed.

File: src/Uno.UI.RuntimeTests/Tests/Extensions/Given_Uri.cs (hunk 7 above)
Comment (aesteves900) [reply src/Uno.UI.RuntimeTests/Tests/Extensions/Given_Uri.cs:13]: It was removed.

**Comment by aesteves900** [comment]
I  put on Given_Uri.cs a #if !WINDOWS_UWP because there is no reference to Uno.Foundation assembly. I tried to add the reference, but it created a lot of compile errors in UWP. I believe there is a reason for the Uno.Foundation assembly is not referenced on UWP.  @carldebilly ?


image

image

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs [hunk 8]
```diff
@@ -42,7 +48,11 @@ public Uri GetCurrentApplicationCallbackUri()
 				this.Log().Warn(message);
 			}
 
-			return new Uri(schemes[0] + WinRTFeatureConfiguration.WebAuthenticationBroker.DefaultCallbackPath);
+			var uri = new Uri(schemes[0] + WinRTFeatureConfiguration.WebAuthenticationBroker.DefaultCallbackPath);
+#if __ANDROID__
+			uri = uri.TrimEndUriSlash();
+#endif
+			return uri;
```
Comment (xperiandri) [inline src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs:55]: Why not optimize it this way then?
```suggestion
            var scheme = schemes[0];
            var hostAndPath = WinRTFeatureConfiguration.WebAuthenticationBroker.DefaultCallbackPath;
            var uriStringBuilder =
                new StringBuilder(scheme.Length + hostNadPath.Length)
                    .Append(scheme);
                    .Append(hostAndPath);
#if __ANDROID__
            if (uriStringBuilder[^1] == '/')
			    uriStringBuilder.Lenght -= 1;
#endif
			return new Uri(uriStringBuilder.ToString());
```

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs (hunk 6 above)
Comment (jeromelaban) [reply src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs:26]: Please do not close the conversation until you've answered the comment or pushed a new branch that addresses the comment.

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs (hunk 8 above)
Comment (aesteves900) [reply src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs:55]: @carldebilly , do you see a needs todo implement in this way for now?

Comment (carldebilly) [reply src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs:55]: That's not required for now.

File: src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs (hunk 6 above)
Comment (aesteves900) [reply src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs:26]: The Android Platform does not handle the last slash in Uri automatically. It is native behavior. Did I answer your question @carldebilly ?

Comment (jeromelaban) [reply src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.cs:26]: @aesteves900 If you do have a link that explains this add it as a comment in the code, otherwise add `The Android Platform does not handle the last slash in Uri automatically.` If you have the error produced when adding the slash, can you add it in the comment as well? It can enhance the troubleshooting experience.

**Review by carldebilly — APPROVED** [review]
